### PR TITLE
Issue 19583 invalid shares

### LINKED ISSUE #19583: Repair step to remove bogus shares from old group memberships
See https://github.com/owncloud/core/issues/19541 and https://github.com/owncloud/core/issues/14384

Write a repair step (or add a case in "RepairInvalidShares") that iterates over all group shares and remove all sub-share entries for users who do not belong to that group any more.

### LINKED ISSUE #20130: Repair step: delete shares where the parent doesn't exist any more
In some rare cases in OC7, bugs might have generated share entries where the parent was deleted but not the children.

### Steps to reproduce:

1. Create two users "user1" and "user2"
1. Login as "user1"
1. Create a folder "test"
1. Share "test" with "user2"
1. Login as "user2"
1. Share "test" with link
1. In the database, delete the parent link entry
1. Go to "Shared with link" section

dirname-error

### Versions
OC 7.0.10, 8.0.6, 8.1.4RC1

A repair step should be written that deletes any `oc_share` entry where the parent does not exist any more.

@rullzer @nickvergessen @schiesbn @MorrisJobke

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Add unit tests

Fix #19583 
Fix #20130

@MorrisJobke @PVince81

### CODE REVIEW

**Comment by nickvergessen** [comment]
@rullzer @MorrisJobke @PVince81 please review

**Comment by MorrisJobke** [comment]
Tested both cases. Works nice :+1:

File: tests/lib/repair/oldgroupmembershipsharestest.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
+<?php
+/**
+ * Copyright (c) 2015 Vincent Petry <pvince81@owncloud.com>
```
Comment (PVince81) [inline tests/lib/repair/oldgroupmembershipsharestest.php]: thanks for the free attribution :wink:

Comment (MorrisJobke) [inline tests/lib/repair/oldgroupmembershipsharestest.php]: @PVince81 We have a repair step for this :P

File: core/command/maintenance/repair.php [hunk 2]
```diff
@@ -49,10 +50,24 @@ public function __construct(\OC\Repair $repair, \OCP\IConfig $config) {
 	protected function configure() {
 		$this
 			->setName('maintenance:repair')
-			->setDescription('repair this installation');
+			->setDescription('repair this installation')
+			->addOption(
+				'include-expensive',
+				null,
```
Comment (PVince81) [inline core/command/maintenance/repair.php]: should this default to true? or do we want to document this specific issue and tell people to add this switch only when they see errors ?

Comment (nickvergessen) [inline core/command/maintenance/repair.php]: No, this may take years (similar to test migration), so it should not default to true.

**Comment by PVince81** [comment]
I'm still ok for this to be merged :+1: 

We can sort out the "expensive repair step" discussion separately.

File: core/command/maintenance/repair.php (hunk 2 above)
Comment (MorrisJobke) [inline core/command/maintenance/repair.php]: Yep. Because otherwise this will be executed during upgrade too and then on big instances can cause trouble.

**Comment by DeepDiver1975** [comment]
can we squash the commits?

**Comment by nickvergessen** [comment]
Done + rebased

**Comment by DeepDiver1975** [comment]
> Done + rebased

looks better - THX

**Comment by MorrisJobke** [comment]
Do we want to backport this to stable8.2 because #19583 was scheduled for 8.2.1?

cc @karlitschek @cmonteroluque

**Comment by karlitschek** [comment]
backport makes sense :+1:

**Pull request merged**